tanjun/checks.py

```diff
@@ -511,6 +569,62 @@ def with_sfw_check(
     return _optional_kwargs(command, SfwCheck(halt_execution=halt_execution, error_message=error_message))
 
 
+def with_cooldown_check(
+    bucket_id: str,
+    /,
+    *,
+    error_message: typing.Optional[str] = "Please wait {cooldown:0.2f} seconds before using this command again",
+    halt_execution: bool = False,
+) -> collections.Callable[[CommandT], CommandT]:
+    """Add cooldown logic to a command through a decorator call.
+
+    .. note::
+        This adds `CooldownCheck` as a check and tanjun.dependencies.CooldownPreExecution`
+        as a pre-execution hook callback to enable cooldown checks.
+
+    Parameters
+    ----------
+    bucket_id : str
+        The bucket id to use for the cooldown check.
+
+    Other Parameters
+    ----------------
+    error_message : typing.Optional[str]
+        The error message to send in response as a command error if the check fails.
+
+        Defaults to "Please wait {cooldown:0.2f} seconds before using this command again" and
+        setting this to `None` will disable the error message allowing the command search to
+        continue.
+    halt_execution : bool
+        Whether this check should raise `tanjun.errors.HaltExecution` to
+        end the execution search when it fails instead of returning `False`.
+
+        Defaults to `False`.
+
+    Notes
+    -----
+    * error_message takes priority over halt_execution.
+    * For more information on how this is used with other parameters see
+      `CallbackReturnT`.
+
+    Returns
+    -------
+    CallbackReturnT[CommandT]
+        The command this check was added to.
+    """
+
+    def decorator(command: CommandT, /) -> CommandT:
+        hooks_ = command.hooks
+        if not hooks_:
+            hooks_ = hooks.AnyHooks()
+            command.set_hooks(hooks_)
+
```

Comment:
Seems like this code would end up racing with other users calling the same bucket, do we want to simplify this by just assuming that if the check is hit and it passes then we should increment the resource or raise a command error in the pre execution hook?